### ISSUE TITLE
baked loading performance bottleneck

### ISSUE THREAD

**Issue by JulianGro** [vircadia/vircadia#1122]
I did some testing with the HQ world and noticed that loading times on http1.1 seem pretty much almost exclusively down to latency.

If I load from a server in Germany (ping 20ms), it takes me ~18 seconds to finish loading.
If I load from the cdn-1 server in the USA (ping 160ms), it takes me ~100 seconds to finish loading.
If I divide the loading times by the number of files (roughly 1000), I end up with a loading time per file that is slightly lower than the ping. 18ms per file on the German server and 100ms per file on the USA server.

The issue here is that the number of assets to load gets multiplied by the number of files oven decides to create. To load one asset we don't just load an fbx file, but we load 1x fst, 1x model, 1x material list, Nx texture file list, Nx texture. `N` being the number of textures used in a model, which is generally around two or three textures for a simple asset.

Since the asset server will have a similar amount of files to serve, a lot of the "asset server is slow" might be coming from this same issue.

There is a good amount of things we can do about this bottleneck:
- One way would be to have the oven just package every asset or every possible request. Currently we only have two different "versions" of an asset, so the oven could just keep two packages per asset.
- Similar to this, we could also have the asset server create packages on the fly. 
E.g. Interface requests model.fst; asset server sends a datastream with model.fbx,material1.json, material2.json, texture1.ktx, texture2.ktx, texture3.ktx, ...
- Another option would be to implement something like http2 push, where rather than waiting for Interface to request a file, we immediately send all files Interface will want. For example when Interface requests model.fst, we would send model.fbx, material1.json and material2.json just after that without waiting for a request.

Personally I am a fan of oven just creating packages out of assets. Ideally we would be smart about the selection of the package and not use an fst file just to give Interface the selection of packages, but have Interface just request the right package.
On http servers you could have some sort of variable in the URL, but on for the asset server we could just implement Interface telling the server once what kind of client it is or maybe even just on every request what kind of asset it wants. E.g. "I want `model` and I am `Mobile`".

Currently the best way to work around this problem seems to be to only bake assets which contain big textures and not bake simple assets to balance the download bottleneck and the baking bottleneck.

**Comment by ctrlaltdavid** [comment]
Perhaps could make the oven pack all an entity's files into a zip file, and download that as one rather than each individual part.

**Comment by JulianGro** [comment]
> Perhaps could make the oven pack all an entity's files into a zip file, and download that as one rather than each individual part.

I would suggest a zstd compressed tar file instead of just zip. It has much higher compression strength and is supposedly a good amount fast as well.
Baked textures compress quite well, so we would also save roughtly 20%-60% of download size.

**Comment by daleglass** [comment]
I've looked a bit into Zstd, and it looks extremely impressive, and with multicore support out of the box in the library.

I think if we're going to break compatibility we should try to get a bunch of improvements in at the same time, to minimize disruption.

**Comment by JulianGro** [comment]
Maybe we should do a smaller change instead. Allow using `.tar.zstd` or just `zstd` URLs, extract the contents of the archive and just load them like you normally would.

For example a baked asset would just be a `.tar.zstd` with all the usual files inside, Interface would just load the fst file and that has a relative path to the other files like normal.

Same could be done with a texture. Just extract the archive and take the texture that us in it. 

Same could be done with pretty much everything you can link to in the entity properties.

There would just need to be some priorities, e.g. if there is multiple files load the fst.
And some warnings if a package didn't include what Interface would expect, e.g. the texture package had multiple files in it. 

I am pretty sure that some `.gz` compressed files can already be loaded. Just fyi.

**Comment by ctrlaltdavid** [comment]
Yes, I don't think we need to or should break compatibility. Rather, add support to optionally package files into a `zstd` for faster download (fewer number of downloads, plus some compression as a bonus).

**Comment by daleglass** [comment]
I looked and there doesn't seem to be .tar handling code in Qt, so we'd need an additional dependency for that.

**Comment by daleglass** [comment]
Some additional info: We use libquazip, which provides a Qt-style interface to .zip files. Quazip itself is based on minizip for the actual compression code, and minizip has a fork with zstd support:

https://github.com/zlib-ng/minizip-ng

So that could be another possibility, perhaps quazip can be made to work with the fork.

**Comment by JulianGro** [comment]
Apparently you are not the first person to think that:
https://github.com/stachenov/quazip/issues/19#issuecomment-812042531

**Comment by JulianGro** [comment]
:angry:
